librz/diff/distance.c

```diff
@@ -0,0 +1,105 @@
+// SPDX-FileCopyrightText: 2017 Fangrui Song <i@maskray.me>
+// SPDX-FileCopyrightText: 2016 NikolaiHampton <nikolaih@3583bytesready.net>
+// SPDX-License-Identifier: LGPL-3.0-only
+#include <rz_diff.h>
+#include <rz_util/rz_assert.h>
+
+// Eugene W. Myers' O(ND) diff algorithm
+// Returns edit distance with costs: insertion=1, deletion=1, no substitution
+RZ_API bool rz_diff_myers_distance(RZ_NONNULL const ut8 *a, ut32 la, RZ_NONNULL const ut8 *b, ut32 lb, RZ_NULLABLE ut32 *distance, RZ_NULLABLE double *similarity) {
+	rz_return_val_if_fail(a && b, false);
+
+	const ut32 length = la + lb;
+	const ut8 *ea = a + la, *eb = b + lb;
+	// Strip prefix
+	for (; a < ea && b < eb && *a == *b; a++, b++) {
+	}
+	// Strip suffix
+	for (; a < ea && b < eb && ea[-1] == eb[-1]; ea--, eb--) {
+	}
+	la = ea - a;
+	lb = eb - b;
+	ut32 *v0, *v;
+	st64 m = (st64)la + lb, di = 0, low, high, i, x, y;
+	if (m + 2 > SIZE_MAX / sizeof(st64) || !(v0 = malloc((m + 2) * sizeof(ut32)))) {
+		return false;
+	}
+	v = v0 + lb;
+	v[1] = 0;
+	for (di = 0; di <= m; di++) {
+		low = -di + 2 * RZ_MAX(0, di - (st64)lb);
+		high = di - 2 * RZ_MAX(0, di - (st64)la);
+		for (i = low; i <= high; i += 2) {
+			x = i == -di || (i != di && v[i - 1] < v[i + 1]) ? v[i + 1] : v[i - 1] + 1;
+			y = x - i;
+			while (x < la && y < lb && a[x] == b[y]) {
+				x++;
+				y++;
+			}
+			v[i] = x;
+			if (x == la && y == lb) {
+				goto out;
+			}
+		}
+	}
+
+out:
+	free(v0);
+	//Clean up output on loop exit (purely aesthetic)
+	if (distance) {
+		*distance = di;
+	}
+	if (similarity) {
+		*similarity = length ? 1.0 - (double)di / length : 1.0;
+	}
+	return true;
+}
+
```

Comment:
Same here. While everyone and their dog knows what Levenstein distance is, some parameters explanation might be helpful for the API users.